### ISSUE TITLE
MINE optimizer is flawed

### ISSUE THREAD

**Issue by devrimcavusoglu** [gtegner/mine-pytorch#5]
I've looked into the code for the implementation and saw [this line](https://github.com/gtegner/mine-pytorch/blob/cb30822620f3dbaef612efa246ed0e1cf210d116/mine/models/mine.py#L81). Here, the point is to get unpaired samples from x_i and zm_i which would intentionally be different than (x_i, z_i). However, `torch.randperm()` does not guarantee derangement, see the following code snippet.

```python
import torch

torch.manual_seed(42)
torch.randperm(5)  # Out[1]: tensor([2, 4, 3, 0, 1])
torch.randperm(5)  # Out[2]: tensor([1, 4, 3, 2, 0])
torch.randperm(5)  # Out[3]: tensor([0, 2, 3, 1, 4])
torch.randperm(5)  # Out[4]: tensor([0, 3, 2, 1, 4])
```

As you can see in the 3rd output we have 0th pos being 0, so derangement is not satisfied, and also in the 4th output both 0 and 2 doesn't satisfy derangement. In my own implementation I directly used a naive approach and made the permutation as j = i + 1, and manually replacing the last item as 0, this naive and simple approach guarantees derangement, but the given batch is important here as it may introduce a bias if the given batch is not well randomized. Please correct me if I'm mistaken, or misunderstood the algorithm/implementation.

**Comment by pzdkn** [comment]
@devrimcavusoglu stumbled across your post. I think all we want is that x,z ~ P(x) * P(z). So shuffling one of them breaks the original pairing. In probability, you might still draw the original pairs (The 3rd output, at pos 0) from the product of the marginals, but it becomes increasingly unlikely when you have a large batch size.
